crates/core/src/db/update.rs

```diff
@@ -167,11 +170,30 @@ pub fn schema_updates(
 ///
 /// - `indexes` are never equal, because they only exist in the database
 /// - `sequences` are never equal, because they only exist in the database
-/// - only `constraints` which have set column attributes can be equal (the
-///   module may emit unset attributes)
+/// - `constraints` are delicate:
 ///
-/// Thus, `indexes` and `sequences` are ignored, while `constraints` are
-/// compared sans any unset attributes.
+///   - The number of `constraints` obtained from the module will always be the
+///     same as the number of columns, emitting [`Constraints::unset()`] if no
+///     specific constraint is defined.
+///
+///   - Indexes defined using `#[spacetimedb(index, ..)]` will not have a
+///     corresponding [`Constraints::indexed()`] constraint (it will be unset).
+///     The schema obtained from the database **will** have one such constraint,
+///     however.
+///
```

Comment:
This is because the conversion into `TableSchema` fills that info. 

`TableDef` is meant to be a partially defined schema that is resolved/qualified when converted to `TableSchema`. See `TableSchema::from_def`

Reply:
Yes, is better.

Reply:
The problem with this is that what the module produces is not a `TableSchema`. The module only specifies a `TableDef`, so by some mechanism we need to compare that to the existing schema (or at least the old def) to find out what changed. You'd have to create a fake `TableSchema` for the new module to compare against.

Reply:
The real problem is that we're not being very careful about what is actual database *schema* (the thing which requires data migration if it changes) and our various attributes (autoinc) or optimizations (indexes).

Indexes are not really part of the schema of a table in the precise meaning of the term schema, neither are attributes.